File: api/scrape.js
let chromium, puppeteer;
try {
  chromium = require('@sparticuz/chromium');
  puppeteer = require('puppeteer-core');
} catch (err) {
  console.log('Puppeteer dependencies not available:', err.message);
}

function buildCorsHeaders(originHeader) {
  const allowedOriginsEnv = process.env.CORS_ORIGINS || '*';
  const allowedOrigins = allowedOriginsEnv.split(',').map(o => o.trim()).filter(Boolean);
  const allowAll = allowedOrigins.includes('*');
  const origin = originHeader || '';
  const allowOriginHeader = allowAll ? '*' : (allowedOrigins.includes(origin) ? origin : allowedOrigins[0] || '*');
  return {
    'Access-Control-Allow-Origin': allowOriginHeader,
    'Access-Control-Allow-Methods': 'POST, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type, x-api-key'
  };
}

function normalizeUrl(inputUrl) {
  try {
    if (!inputUrl) return '';
    if (!/^https?:\/\//i.test(inputUrl)) {
      return `https://${inputUrl}`;
    }
    return inputUrl;
  } catch (_) {
    return '';
  }
}

function inferPricingModel(text) {
  const lower = text.toLowerCase();
  if (/contact\s+(sales|us)|talk\s+to\s+sales|request\s+a\s+quote|custom\s+pricing/.test(lower)) return 'Custom';
  if (/(per\s*user|\/user|per\s*seat|\/seat)/.test(lower)) return 'Per-User';
  if (/(per\s*(api|gb|request|usage|unit)|\busage\b|pay\s+as\s+you\s+go)/.test(lower)) return 'Usage-Based';
  if (/(free|\$?0\b)/.test(lower)) return 'Freemium';
  return 'Tiered';
}

function parseBillingCycle(text) {
  const lower = text.toLowerCase();
  if (/month|mo\b/.test(lower)) return 'monthly';
  if (/year|annual|yr\b/.test(lower)) return 'annually';
  if (/(per\s*user|\/user|per\s*seat|\/seat)/.test(lower)) return 'per user';
  return 'N/A';
}

export default async function handler(req, res) {
  const corsHeaders = buildCorsHeaders(req.headers.origin);

  // CORS preflight
  if (req.method === 'OPTIONS') {
    return res.status(200).set(corsHeaders).send('');
  }

  if (req.method !== 'POST') {
    return res.status(405).set({ 'Content-Type': 'application/json', ...corsHeaders }).json({ error: 'Method not allowed' });
  }

  // API key auth
  const apiKeyHeader = req.headers['x-api-key'];
  const expectedKey = process.env.SCRAPER_API_KEY || '';
  if (!expectedKey || apiKeyHeader !== expectedKey) {
    return res.status(401).set({ 'Content-Type': 'application/json', ...corsHeaders }).json({ error: 'Unauthorized' });
  }

  try {
    const { url } = typeof req.body === 'string' ? JSON.parse(req.body || '{}') : (req.body || {});
    const targetUrl = normalizeUrl(url);
    if (!targetUrl) {
      return res.status(400).set({ 'Content-Type': 'application/json', ...corsHeaders }).json({ error: 'Invalid or missing url' });
    }

    // Check if puppeteer is available
    if (!chromium || !puppeteer) {
      return res.status(500).set({ 'Content-Type': 'application/json', ...corsHeaders }).json({ 
        error: 'Puppeteer dependencies not available. Try /api/scrape-simple for basic scraping.' 
      });
    }

    // Configure Chromium for serverless
    chromium.setHeadlessMode = true;
    chromium.setGraphicsMode = false;

    const launchOptions = {
      args: chromium.args,
      defaultViewport: chromium.defaultViewport,
      executablePath: await chromium.executablePath(),
      headless: chromium.headless,
      ignoreHTTPSErrors: true,
    };

    const browser = await puppeteer.launch(launchOptions);
    try {
      const page = await browser.newPage();
      await page.setUserAgent('Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124 Safari/537.36');
      await page.setRequestInterception(true);
      page.on('request', (request) => {
        const resourceType = request.resourceType();
        if (['image', 'media', 'font'].includes(resourceType)) {
          request.abort();
        } else {
          request.continue();
        }
      });

      // Navigate with tight timeout to fit serverless limits
      try {
        await page.goto(targetUrl, { waitUntil: ['domcontentloaded', 'networkidle2'], timeout: 8000 });
      } catch (_) {
        // Fallback to at least DOM ready
        try { await page.goto(targetUrl, { waitUntil: 'domcontentloaded', timeout: 8000 }); } catch (_) {}
      }

      // Small grace period for late content
      await page.waitForTimeout(300);

      const results = await Promise.race([
        page.evaluate(() => {
          function getText(el) { return (el && el.textContent || '').replace(/\s+/g, ' ').trim(); }
          const currencyRe = /(?:\$|€|£)\s?\d{1,3}(?:[\,\.]\d{3})*(?:[\.,]\d{2})?|\bfree\b|contact\s+(?:sales|us)|request\s+a\s+quote/i;

          // Find nodes with price-like text
          const walker = document.createTreeWalker(document.body, NodeFilter.SHOW_ELEMENT);
          const priceNodes = [];
          while (walker.nextNode()) {
            const node = walker.currentNode;
            const text = getText(node);
            if (!text) continue;
            if (currencyRe.test(text)) {
              priceNodes.push(node);
            }
            if (priceNodes.length > 120) break; // cap for performance
          }

          // For each price node, find a reasonable plan card ancestor
          function findCard(node) {
            let cur = node;
            let depth = 0;
            while (cur && depth < 6) {
              const text = getText(cur).toLowerCase();
              const hasList = cur.querySelectorAll('li').length >= 2;
              const hasHeading = cur.querySelector('h1,h2,h3,h4,h5,h6,[class*="plan" i],[class*="tier" i],[class*="title" i]');
              const containsMultiplePrices = (text.match(/\$|€|£/g) || []).length <= 6; // avoid grabbing whole page
              const hasColumns = getComputedStyle(cur).display.includes('grid') || getComputedStyle(cur).display.includes('flex');
              if ((hasHeading || hasList || hasColumns) && containsMultiplePrices) {
                return cur;
              }
              cur = cur.parentElement;
              depth += 1;
            }
            return node;
          }

          const cards = new Set();
          const planCards = [];
          for (const pn of priceNodes) {
            const card = findCard(pn);
            if (!card) continue;
            if (cards.has(card)) continue;
            cards.add(card);
            planCards.push(card);
            if (planCards.length > 12) break; // cap
          }

          function extractFromCard(card) {
            const text = getText(card);
            // Price
            let price = '';
            const priceEl = card.querySelector('[class*="price" i], [data-test*="price" i]');
            price = priceEl ? getText(priceEl) : '';
            if (!price || !/(\$|€|£|free|contact|quote)/i.test(price)) {
              const m = text.match(currencyRe);
              if (m) price = m[0];
            }

            // Plan name
            let planName = '';
            const heading = card.querySelector('h1,h2,h3,h4,h5,h6,[class*="plan" i],[class*="tier" i],[class*="name" i],[class*="title" i]');
            if (heading) planName = getText(heading);
            if (!planName) {
              // Try previous sibling heading
              let prev = card.previousElementSibling;
              if (prev) {
                const h = prev.querySelector('h1,h2,h3,h4,h5,h6');
                if (h) planName = getText(h);
              }
            }

            // Features
            let features = Array.from(card.querySelectorAll('li'))
              .map(li => getText(li))
              .filter(Boolean)
              .slice(0, 20);

            // Billing cycle (from card text or price)
            let billingText = '';
            const cycleCandidates = [
              card.querySelector('[class*="billing" i]'),
              card.querySelector('[class*="cycle" i]'),
              card.querySelector('[class*="period" i]'),
              card.querySelector('[class*="per" i]')
            ].filter(Boolean);
            if (cycleCandidates.length) billingText = getText(cycleCandidates[0]);
            const billing_cycle = (() => {
              const t = (billingText || price || text).toLowerCase();
              if (/month|mo\b/.test(t)) return 'monthly';
              if (/year|annual|yr\b/.test(t)) return 'annually';
              if (/(per\s*user|\/user|per\s*seat|\/seat)/.test(t)) return 'per user';
              return 'N/A';
            })();

            // Pricing model
            const pricing_model = (() => {
              const t = (text + ' ' + price).toLowerCase();
              if (/contact\s+(sales|us)|talk\s+to\s+sales|request\s+a\s+quote|custom\s+pricing/.test(t)) return 'Custom';
              if (/(per\s*user|\/user|per\s*seat|\/seat)/.test(t)) return 'Per-User';
              if (/(per\s*(api|gb|request|usage|unit)|\busage\b|pay\s+as\s+you\s+go)/.test(t)) return 'Usage-Based';
              if (/(^|\s)(free|\$?0\b)/.test(t)) return 'Freemium';
              return 'Tiered';
            })();

            return { plan_name: planName || '', price: price || '', pricing_model, features, billing_cycle };
          }

          const extracted = planCards.map(extractFromCard)
            .filter(p => p.price || p.plan_name)
            .slice(0, 8);

          // Deduplicate by plan name + price
          const seen = new Set();
          const unique = [];
          for (const p of extracted) {
            const key = (p.plan_name + '|' + p.price).toLowerCase();
            if (seen.has(key)) continue;
            seen.add(key);
            unique.push(p);
          }

          return unique;
        }),
        new Promise((_, reject) => setTimeout(() => reject(new Error('Extraction timeout')), 2500))
      ]);

      return res.status(200).set({ 'Content-Type': 'application/json', ...corsHeaders }).json(Array.isArray(results) ? results : []);
    } finally {
      await browser.close().catch(() => {});
    }
  } catch (err) {
    return res.status(500).set({ 'Content-Type': 'application/json', ...corsHeaders }).json({ error: String(err && err.message || err) });
  }
}

File: api/scrape.py
import json
import requests
from bs4 import BeautifulSoup
import re
from urllib.parse import urljoin, urlparse

def handler(request, context):
    # Handle CORS preflight
    if request.get('method') == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': {
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Methods': 'POST, OPTIONS',
                'Access-Control-Allow-Headers': 'Content-Type, x-api-key'
            },
            'body': ''
        }
    
    if request.get('method') != 'POST':
        return {
            'statusCode': 405,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({'error': 'Method not allowed'})
        }
    
    # API key authentication
    api_key_header = request.get('headers', {}).get('x-api-key')
    expected_key = context.get('SCRAPER_API_KEY', 'test-key')  # Default for testing
    if not api_key_header or api_key_header != expected_key:
        return {
            'statusCode': 401,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({'error': 'Unauthorized'})
        }
    
    try:
        # Parse request body
        body = request.get('body', '{}')
        if isinstance(body, str):
            data = json.loads(body)
        else:
            data = body
        
        url = data.get('url', '').strip()
        if not url:
            return {
                'statusCode': 400,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*'
                },
                'body': json.dumps({'error': 'Missing url parameter'})
            }
        
        # Normalize URL
        if not url.startswith(('http://', 'https://')):
            url = 'https://' + url
        
        # Fetch the page
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        }
        
        response = requests.get(url, headers=headers, timeout=10)
        response.raise_for_status()
        
        # Parse HTML
        soup = BeautifulSoup(response.content, 'html.parser')
        
        # Extract pricing plans
        plans = extract_pricing_plans(soup, url)
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps(plans)
        }
        
    except requests.RequestException as e:
        return {
            'statusCode': 500,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({'error': f'Failed to fetch URL: {str(e)}'})
        }
    except Exception as e:
        return {
            'statusCode': 500,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps({'error': f'Processing error: {str(e)}'})
        }

def extract_pricing_plans(soup, base_url):
    """Extract pricing plans from HTML using heuristics"""
    plans = []
    
    # Find potential pricing containers
    pricing_selectors = [
        '[class*="pricing" i]',
        '[class*="plan" i]',
        '[class*="tier" i]',
        '[class*="package" i]',
        '[class*="price" i]',
        '[id*="pricing" i]',
        '[id*="plan" i]'
    ]
    
    pricing_containers = []
    for selector in pricing_selectors:
        containers = soup.select(selector)
        pricing_containers.extend(containers)
    
    # Also look for common pricing patterns
    currency_regex = r'[\$€£¥]\s*\d{1,3}(?:,\d{3})*(?:\.\d{2})?'
    price_elements = soup.find_all(text=re.compile(currency_regex))
    
    for elem in price_elements:
        parent = elem.parent
        if parent:
            pricing_containers.append(parent)
    
    # Remove duplicates and get unique containers
    unique_containers = list(set(pricing_containers))
    
    # Extract plans from containers
    for container in unique_containers[:10]:  # Limit to avoid too many results
        plan = extract_plan_from_container(container)
        if plan and plan['price']:  # Only add if we found a price
            plans.append(plan)
    
    # Deduplicate plans
    seen_plans = set()
    unique_plans = []
    for plan in plans:
        plan_key = f"{plan['plan_name']}|{plan['price']}"
        if plan_key not in seen_plans:
            seen_plans.add(plan_key)
            unique_plans.append(plan)
    
    return unique_plans[:8]  # Return max 8 plans

def extract_plan_from_container(container):
    """Extract plan details from a container element"""
    text = container.get_text() if hasattr(container, 'get_text') else str(container)
    
    # Extract price
    price_regex = r'[\$€£¥]\s*\d{1,3}(?:[,\.]\d{3})*(?:[,\.]\d{2})?(?:/\w+)?'
    price_match = re.search(price_regex, text)
    price = price_match.group(0) if price_match else ''
    
    # Handle special pricing
    if re.search(r'contact\s+(sales|us)|custom|enterprise', text, re.IGNORECASE):
        if not price:
            price = 'Contact Sales'
    elif re.search(r'\bfree\b', text, re.IGNORECASE) and not price:
        price = 'Free'
    
    # Extract plan name
    plan_name = ''
    if hasattr(container, 'find'):
        # Look for headings
        heading = container.find(['h1', 'h2', 'h3', 'h4', 'h5', 'h6'])
        if heading:
            plan_name = heading.get_text().strip()
    
    if not plan_name:
        # Try common plan names
        plan_patterns = ['basic', 'starter', 'pro', 'premium', 'enterprise', 'business', 'free', 'plus']
        for pattern in plan_patterns:
            if re.search(rf'\b{pattern}\b', text, re.IGNORECASE):
                plan_name = pattern.title()
                break
    
    # Extract features
    features = []
    if hasattr(container, 'find_all'):
        # Look for list items
        li_elements = container.find_all('li')
        for li in li_elements[:10]:  # Limit features
            feature_text = li.get_text().strip()
            if feature_text and len(feature_text) < 100:
                features.append(feature_text)
    
    # Determine billing cycle
    billing_cycle = 'N/A'
    if re.search(r'/month|monthly|per month', text, re.IGNORECASE):
        billing_cycle = 'monthly'
    elif re.search(r'/year|yearly|annually|per year', text, re.IGNORECASE):
        billing_cycle = 'annually'
    elif re.search(r'per user|/user|per seat|/seat', text, re.IGNORECASE):
        billing_cycle = 'per user'
    
    # Determine pricing model
    pricing_model = 'Tiered'  # Default
    if re.search(r'contact\s+(sales|us)|custom|enterprise', text, re.IGNORECASE):
        pricing_model = 'Custom'
    elif re.search(r'per user|/user|per seat|/seat', text, re.IGNORECASE):
        pricing_model = 'Per-User'
    elif re.search(r'usage|pay as you go|api call|per request', text, re.IGNORECASE):
        pricing_model = 'Usage-Based'
    elif re.search(r'\bfree\b|\$0', text, re.IGNORECASE):
        pricing_model = 'Freemium'
    
    return {
        'plan_name': plan_name or 'Unknown Plan',
        'price': price,
        'pricing_model': pricing_model,
        'features': features,
        'billing_cycle': billing_cycle
    }